contracts/DefaultStrategy.sol

```diff
@@ -0,0 +1,517 @@
+     */
+    error MaxGroupsVotedForReached();
+
+    /**
+     * @notice Initialize the contract with registry and owner.
+     * @param _registry The address of the Celo registry.
+     * @param _owner The address of the contract owner.
+     * @param _manager The address of the Manager contract.
+     */
+    function initialize(
+        address _registry,
+        address _owner,
+        address _manager
+    ) external initializer {
+        _transferOwnership(_owner);
+        __UsingRegistry_init(_registry);
+        __Managed_init(_manager);
+    }
+
+    /**
+     * @notice Set this contract's dependencies in the StakedCelo system.
+     * @param _account The address of the Account contract.
+     * @param _groupHealth The address of the GroupHealth contract.
+     * @param _specificGroupStrategy The address of the SpecificGroupStrategy contract.
+     */
+    function setDependencies(
+        address _account,
+        address _groupHealth,
+        address _specificGroupStrategy
+    ) external onlyOwner {
+        require(_account != address(0), "Account null");
+        require(_groupHealth != address(0), "GroupHealth null");
+        require(_specificGroupStrategy != address(0), "SpecificGroupStrategy null");
+
+        groupHealth = IGroupHealth(_groupHealth);
+        specificGroupStrategy = ISpecificGroupStrategy(_specificGroupStrategy);
+        account = IAccount(_account);
+    }
+
+    /**
+     * @notice Distributes votes by computing the number of votes each active
+     * group should receive, then calling out to `Account.scheduleVotes`.
+     * @param votes The amount of votes to distribute.
+     * @dev The vote distribution strategy is to try and have each validator
+     * group to be receiving the same amount of votes from the system. If a
+     * group already has more votes than the average of the total available
+     * votes it will not be voted for, and instead we'll try to evenly
+     * distribute between the remaining groups.
+     * @dev Election.sol sets a dynamic limit on the number of votes receivable
+     * by a group, based on the group's size, the total amount of Locked
+     * CELO, and the total number of electable validators. We don't want to
+     * schedule votes for a group when the amount would exceed this threshold.
+     * `getVotableGroups` below selects those groups that could receive the
+     * entire `votes` amount, and filters out the rest. This is a heuristic:
+     * when distributing votes evenly, the group might receive less than
+     * `votes`, and the total amount could end up being under the limit.
+     * However, doing an exact computation would be both complex and cost a lot
+     * of additional gas, hence the heuristic. If indeed all groups are close to
+     * their voting limit, causing a larger deposit to revert with
+     * NoVotableGroups, despite there still being some room for deposits, this
+     * can be worked around by sending a few smaller deposits.
```

Comment:
Ah, I guess one thing to notice here: now this can also fail when doing a transfer. In particular, to/from smart contracts. So in theory there could be a situation where a smart contract that holds an stCELO balance could have some of its functionality frozen.
On that topic, do we check if new votes scheduled for a specific group go over the Election.sol vote limit? I guess there's something to worry about in both cases:
* If we don't check, there could be "useless" stCELO in the system, that is not actually used in voting. Also, couldn't `activateAndVote` break, since it always votes with the entire amount scheduled?
* If we do check, again, transfers to specific accounts can end up getting frozen.

Rough idea of a potential solution that just came to mind: have an "overflow strategy" that can't normally be selected. When votes can't be cast normally, we keep track of how many votes should be cast for the selected strategy, but allow for real votes to diverge - the votes end up actually cast according to this overflow strategy. At any point, if an "overflowed" group has capacity again, votes can be rebalanced back from the overflow strategy to a group that should have received those votes.

Reply:
We currently don't check it, but the vote transaction itself will revert. 

I will try to implement the overflow strategy

Reply:
We could even disallow voting for such specific group when limit is reached and account is trying to vote again. For default we should have overflow strategy

Reply:
Also Im wondering whether one overflow strategy would be enough or whether we should generalise it and have array of overflow strategies

Reply:
An alternative could be:

- For specific validator group voting:
> Revert deposits
> Allow transfers and use default strategy as overflow strategy
- Change default strategy to Round robin

What do you think ?

Reply:
As discussed - this will be addressed in separate task